### PR TITLE
Add an argument to be disable proguard code generating

### PR DESCRIPTION
**Pull request by SeongUgJung**
This is to make an option to disable proguard rule generating.
I put an option argument : `moshi.enabledProguardGenerated`. 
Once it sets `false`, it will skip to generate proguard rule.

Issue : https://github.com/square/moshi/issues/1351

### CODE REVIEW

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt [hunk 1]
```diff
@@ -59,6 +59,13 @@ public class JsonClassCodegenProcessor : AbstractProcessor() {
      *   * `"javax.annotation.Generated"` (JRE <9)
      */
     public const val OPTION_GENERATED: String = "moshi.generated"
+
+    /**
+     * This annotation processing argument can disable proguard rule generating.
+     * Normally, this is not recommended unless end-users build their own JsonAdapter look-up tool.
+     * This is enabled by default
+     */
+    public const val OPTION_PROGUARD_CODE_GENERATED: String = "moshi.enabledProguardGenerated"
```
Comment (ZacSweers) [inline kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt:68]: Let's call this `OPTION_ENABLE_PROGUARD_RULE_GENERATION` and the value `mocha.enableProguardRuleGeneration`

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt [hunk 2]
```diff
@@ -72,6 +79,7 @@ public class JsonClassCodegenProcessor : AbstractProcessor() {
   private lateinit var cachedClassInspector: MoshiCachedClassInspector
   private val annotation = JsonClass::class.java
   private var generatedType: ClassName? = null
+  private var generatedProguard: Boolean = true
```
Comment (ZacSweers) [inline kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt:82]: Rename this to `generateProguardRules`

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt [hunk 3]
```diff
@@ -134,7 +145,16 @@ public class JsonClassCodegenProcessor : AbstractProcessor() {
           }
 
         preparedAdapter.spec.writeTo(filer)
-        preparedAdapter.proguardConfig?.writeTo(filer, type)
+        if (generatedProguard) {
+          preparedAdapter.proguardConfig?.writeTo(filer, type)
+        } else {
+          messager.printMessage(
```
Comment (ZacSweers) [inline kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt:151]: Let's not print a noisy warning for this and trust that people specifying the option to disable it know what they're doing 👍

File: kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt [hunk 4]
```diff
@@ -355,6 +355,27 @@ class JsonClassCodegenProcessorTest {
...
+          import com.squareup.moshi.JsonClass
+
+          @JsonClass(generateAdapter = true)
+          data class Foo(val a: Int)
+          """
+      )
+    ).apply {
+      kaptArgs[JsonClassCodegenProcessor.OPTION_PROGUARD_CODE_GENERATED] = "false"
+    }.compile()
+    assertThat(result.messages).contains(
```
Comment (ZacSweers) [inline kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt:373]: After removing the warning, change this test to just assert that no rules files were generated

**Review by ZacSweers — COMMENTED** [review]
Thanks! I like the overall structure, just want to see some wording changes and remove the printed warning. Keeping review consolidated here, after this PR is done we can bring the MoshiX PR up to parity with it

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt (hunk 1 above)
Comment (seongug-jung) [reply kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt]: updated on bf273da

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt (hunk 2 above)
Comment (seongug-jung) [reply kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt]: updated on bf273da

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt (hunk 3 above)
Comment (seongug-jung) [reply kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt]: removed printing

File: kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt (hunk 4 above)
Comment (seongug-jung) [reply kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt]: by removing, I replaced this with result.generatedFiles assertion. bf273da

**Comment by SeongUgJung** [comment]
🤦 Please don't confuse other person : @seongug-jung. Preciously It is same person but office account

File: kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt [hunk 5]
```diff
@@ -59,6 +59,13 @@ public class JsonClassCodegenProcessor : AbstractProcessor() {
      *   * `"javax.annotation.Generated"` (JRE <9)
      */
     public const val OPTION_GENERATED: String = "moshi.generated"
+
+    /**
+     * This annotation processing argument can disable proguard rule generating.
```
Comment (ZacSweers) [inline kotlin/codegen/src/main/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessor.kt:64]:
```suggestion
     * This annotation processing argument disables proguard rule generation.
```

File: kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt [hunk 6]
```diff
@@ -355,6 +355,23 @@ class JsonClassCodegenProcessorTest {
...
+          import com.squareup.moshi.JsonClass
+
+          @JsonClass(generateAdapter = true)
+          data class Foo(val a: Int)
+          """
+      )
+    ).apply {
+      kaptArgs[JsonClassCodegenProcessor.OPTION_ENABLE_PROGUARD_RULE_GENERATION] = "false"
+    }.compile()
+    assertThat(result.generatedFiles.all { !it.endsWith(".pro") }).isFalse()
```
Comment (ZacSweers) [inline kotlin/codegen/src/test/java/com/squareup/moshi/kotlin/codegen/JsonClassCodegenProcessorTest.kt:373]:
```suggestion
    assertThat(result.generatedFiles.filter { it.endsWith(".pro") }).isEmpty()
```

**Comment by ZacSweers** [comment]
Apologies for the delay on getting to this, thanks!

**Pull request merged**